File: src/simple_publisher_subscriber_py/simple_publisher_subscriber_py/publisher_member_function.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import time
import sys


class MinimalPublisher(Node):

    def __init__(self, exec_time=300, sleep_time=0.25):
        super().__init__('minimal_publisher')
        self.publisher_ = self.create_publisher(String, 'topic', 10)
        timer_period = sleep_time
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.i = 0
        self.start_time = time.time()

        self.exec_time = exec_time
        self.sleep_time = sleep_time

    def timer_callback(self):
        current_time = time.time()
        elapsed_time = current_time - self.start_time

        if elapsed_time >= self.exec_time:
            self.get_logger().info('Tempo limite atingido, encerrando o nó [pub].')
            sys.exit(0)

        msg = String()
        msg.data = 'Hello World: %d' % self.i
        self.publisher_.publish(msg)
        self.get_logger().info('Publishing: "%s"' % msg.data)
        self.i += 1

        time.sleep(self.sleep_time)


def main(args=None):
    rclpy.init(args=args)

    if len(sys.argv) >= 3:
        exec_time = float(sys.argv[1])
        sleep_time = float(sys.argv[2])
    else:
        exec_time = 300
        sleep_time = 0.25

    minimal_publisher = MinimalPublisher(exec_time=exec_time, sleep_time=sleep_time)

    try:
        rclpy.spin(minimal_publisher)
    except KeyboardInterrupt:
        pass

    if rclpy.ok():
        minimal_publisher.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/simple_publisher_subscriber_cpp/src/publisher_member_function.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <thread>
#include <cstdlib>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node
{
public:
  MinimalPublisher(double execution_duration, double sleep_duration)
  : Node("minimal_publisher"), count_(0), execution_duration_(execution_duration), sleep_duration_(sleep_duration)
  {
    start_time_ = this->now();
    
    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
    
    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(static_cast<int>(sleep_duration_ * 1000)),
      std::bind(&MinimalPublisher::timer_callback, this));
  }

private:
  void timer_callback()
  {
    auto current_time = this->now();
    auto duration = current_time - start_time_;
    double elapsed_seconds = duration.seconds();

    if (elapsed_seconds >= execution_duration_) {
      RCLCPP_INFO(this->get_logger(), "Tempo limite atingido, encerrando o nó.");
      exit(0);
    }

    auto message = std_msgs::msg::String();
    message.data = "Hello, world! " + std::to_string(count_++);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
  }

  rclcpp::Time start_time_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  size_t count_;
  double execution_duration_;
  double sleep_duration_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);

  double execution_duration = 300.0;
  double sleep_duration = 0.25;

  if (argc >= 3) {
    execution_duration = std::atof(argv[1]);
    sleep_duration = std::atof(argv[2]);
  } else {
    RCLCPP_WARN(rclcpp::get_logger("rclcpp"), "Usando valores padrão: execution_duration = 300.0, sleep_duration = 0.25");
  }

  auto node = std::make_shared<MinimalPublisher>(execution_duration, sleep_duration);
  rclcpp::spin(node);

  rclcpp::shutdown();

  return 0;
}
